### ISSUE TITLE
Serializing to and deserializing from collections of JsonElement disregards JsonClassDiscriminator  (1.5.0-RC, experimental, kotlin 1.8.0)

### ISSUE THREAD

**Issue by bjerga** [Kotlin/kotlinx.serialization#2179]
**Describe the bug**

Serializing and deserializing a sealed class annotated with `JsonClassDiscriminator` fails to include the custom discriminator for collections of `JsonElement`.

I've tried to illustrate which Json-methods work as expected and which don't in the following tables:

| 1.4.1                 | Single value       | Collection         |
|-----------------------|--------------------|--------------------|
| encodeToString        | :heavy_check_mark: | :heavy_check_mark: |
| decodeFromString      | :heavy_check_mark:                | :heavy_check_mark: |
| encodeToJsonElement   | :heavy_check_mark: | :heavy_check_mark: |
| decodeFromJsonElement | :heavy_check_mark:                | :heavy_check_mark: |

| 1.5.0-RC              | Single value       | Collection         |
|-----------------------|--------------------|--------------------|
| encodeToString        | :heavy_check_mark: | :heavy_check_mark: |
| decodeFromString      | :heavy_check_mark:                | :heavy_check_mark: |
| encodeToJsonElement   | :heavy_check_mark: | :x:                |
| decodeFromJsonElement | :heavy_check_mark:                | :x:                |

**To Reproduce**

I'm using `List` in the example below, but the issue is also present with `Set` and `Map`.

```kotlin
@Serializable
@OptIn(ExperimentalSerializationApi::class)
@JsonClassDiscriminator("hobbit")
sealed class Hobbit {
    @Serializable
    @SerialName("Frodo")
    object Frodo : Hobbit()

    @Serializable
    @SerialName("Samwise")
    data class Samwise(val quality: String) : Hobbit()
}

val hobbits = listOf<Hobbit>(Hobbit.Frodo, Hobbit.Samwise("brave"))

// Will erroneously serialize to """[{"type":"Frodo"},{"type":"Samwise","quality":"brave"}]"""
val hobbitsSerialized = hobbits.let(Json::encodeToJsonElement).toString()

// Fails with message "Polymorphic serializer was not found for missing class discriminator ('null')"
val hobbitsDeserialized = """[{"hobbit":"Frodo"},{"hobbit":"Samwise","quality":"brave"}]""".let(Json::parseToJsonElement).let { Json.decodeFromJsonElement<List<Hobbit>>(it) }
```

**Expected behavior**
 
I expect `hobbits.let(Json::encodeToJsonElement).toString()` to serialize to `"""[{"hobbit":"Frodo"},{"hobbit":"Samwise","quality":"brave"}]"""`.

I expect `"""[{"hobbit":"Frodo"},{"hobbit":"Samwise","quality":"brave"}]""".let(Json::parseToJsonElement).let { Json.decodeFromJsonElement<List<Hobbit>>(it) }` to correctly deserialize into a list like `[Frodo, Samwise(quality="brave")]`.

**Environment**
 - Kotlin version: 1.8.0
 - Library version: 1.5.0-RC
 - Kotlin platforms: JVM
 - Gradle version: 7.5.1
 - OS version: Windows 10 Enterprise 21H2 19044.2486

**Comment by sandwwraith** [comment]
It looks like the error is related to the new intrinsics functionality (#1348). There are several workarounds:
1. Pass `Hobbit.serializer().list` explicitly to `encodeToJsonElement()`
2. Add `["-P", "plugin:org.jetbrains.kotlinx.serialization:disableIntrinsic=true"]` to `freeCompilerArgs` in the buildscript

**Comment by sandwwraith** [comment]
@bjerga It's fixed in Kotlin master and 1.8.20 branches
